### PR TITLE
Wait for pipeline state in Data Fusion operators

### LINKED ISSUE #8673: Data Fusion Hook Start pipeline will succeed before pipeline is in RUNNING state
(re-opening of #8672 with appropriate template)
CC: @turbaszek 
**Apache Airflow version**: master

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`):

**Environment**: Composer (this is not unique to composer it's a code logic issue)

- **Cloud provider or hardware configuration**: gcp
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:
[Here](https://github.com/apache/airflow/blob/4421f011eeec2d1022a39933e27f530fb9f9c1b1/airflow/providers/google/cloud/hooks/datafusion.py#L392) the `start_pipeline` method of data fusion hook will succeed if they get a 200 from the CDAP API.
This is a misleading success signal as this indicates at best that this pipeline entered the PENDING state. However start pipeline should not succeed until the pipline has reached the RUNNING state.
Note the Happy path is PENDING > STARTING> RUNNING ([ProgramStatus](https://github.com/cdapio/cdap/blob/1d62163faaecb5b888f4bccd0fcf4a8d27bbd549/cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunStatus.java)) Many CDAP pipelines using Dataproc Provisioner spend a significant amount of time in the STARTING state because they  also have tick through the various [ProgramRunClusterStatus](https://github.com/cdapio/cdap/blob/1d62163faaecb5b888f4bccd0fcf4a8d27bbd549/cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunClusterStatus.java) for provisioning the dataproc cluster.

**What you expected to happen**:
start_pipeline and the associated operator should not succeed until pipeline is started (enters RUNNING state). 

**How to reproduce it**:
This is a code logic issue and will be reproduced by any use of this method.
If you want to demonstrate why this is problematic.
- Run a dag that runs [CloudDataFusionStartPipelineOperator](https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/operators/datafusion.py#L607)
- Do any of the following: 
    - Hop over to Data Fusion UI and manually stop the pipeline before it enters the running state
    - Manually delete the dataproc cluster before it finishes provisioning
    - Use a Compute Profile that tries to provision an illegal dataproc cluster (e.g. due to permissions issue where CDF SA doesn't have sufficient permission to create cluster in another project)  
- observe that CloudDataFusionStartPipelineOperator task succeeds (even though pipeline never really started).

We could catch things like this in the future by adding a system test that uses CloudDataFusionStartPipelineOperator (and the test immediately kills it using the stop endpoint) and asserts that the task fails.

**Anything else we need to know**:
Unfortunately making the start call to CDAP does not return a run_id to poll for state.

This hook could work around this by adding a special runtime arg e.g. `__faux_airflow_id__` which can be used to "look up" the real run id by this special runtime arg. the value of this runtime arg could be the dag_run_id or something. If using this workaround or CDAP API can return run id, then a more useful operator than start pipeline would be one that actually waits til the job reaches a success state (much like the existing dataflow and dataproc operators).

Example in golang for terraform provider resource that manages a streaming pipeline.
[Creating with faux id](https://github.com/GoogleCloudPlatform/terraform-provider-cdap/blob/master/cdap/resource_streaming_program_run.go#L108) 

And [looking up the real CDAP run id by faux id](https://github.com/GoogleCloudPlatform/terraform-provider-cdap/blob/master/cdap/resource_streaming_program_run.go#L216)

### PR DESCRIPTION
**Pull request by turbaszek**
Closes: #8673

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: It looks like this `_get_pipeline` gets called every iteration of the polling while loop.
this means on each poll we loop through all the program runs (which may be numerous).

We should only run this loop to "look up program run by faux id" once to get the "real" CDAP job ID.
This real job ID can be used in polling to get this program run directly with a GET request.

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: typo
```suggestion
        Polls pipeline state and raises an exception if the state is one of
```

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: nit: IMO this should provide context link to relevant  CDAP bug https://issues.cask.co/browse/CDAP-7641
This way once that is closed we can remember to simplify this.
```suggestion
        # Unfortunately making the start call to CDAP does not return a run_id to poll for state.
        # https://issues.cask.co/browse/CDAP-7641
```

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: It would be easier for the user to understand what started a job (in the CDAP UI) if this contained the airflow task id or dag run id.

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: nit: if `monotonic() - start_time == timeout` (or `monotonic - start_time - timeout < 30`) in  above if statement we should not wait 30 seconds to raise exception.

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: While true seems infinite loop prone. Any reason to not merge this with the below if?
```python3
    while monotonic - start_time > timepout:
        ...
    raise AirflowException(....)
```

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: should `success_states` have a reasonable default similar to failure states?

File: airflow/providers/google/cloud/operators/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/operators/datafusion.py]: This operator is name `CloudDataFusionStartPipelineOperator` (key word start).
IMO that means `PipelineStates.RUNNING` should be a success state by default.
However, it might be worth allowing the user to optionally control these success states to span more use cases.

**Review by jaketf — CHANGES_REQUESTED** [review]

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/datafusion.py:446]: @jaketf do I correctly understand that state of program can be acceses under:
```
https://xxx-dot-eun1.datafusion.googleusercontent.com/api/v3/namespaces/default/apps/airflow_test/workflows/DataPipelineWorkflow/runs/5d1a277b-17ae-4cab-b3a6-5d23a2f8f672
```
at least that's how it seems to be done here
https://github.com/GoogleCloudPlatform/terraform-provider-cdap/blob/e16a2c5188bf2299482d5c8261ca01f27cd19a44/cdap/resource_streaming_program_run.go#L244

The problem is that I get 404, any ideas?

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/datafusion.py]: Then the id will have to be generated in the operator. From hook level we are unable (in a simple way) access this information.

File: airflow/providers/google/cloud/operators/datafusion.py
Comment (turbaszek) [reply airflow/providers/google/cloud/operators/datafusion.py]: I've added `success_states` as an optional argument. If not provided the operator will wait for operation to be `RUNNING`

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [reply airflow/providers/google/cloud/hooks/datafusion.py]: Could the hook parameterize this faux id to let the operator do this?

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: We can give a little more info here
```suggestion
            f"Unable to retrieve run id of `{pipeline_name}` pipeline with runtime arg {job_id_key}={faux_pipeline_id}."
```

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: nit we already have these defined in module a constant
```suggestion
            success_states=SUCCESS_STATES,
```

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: How often do you notice the first request failing? If >50% then we expect the program run doesn't show up for some time (seconds) and we can expect first iteration to fail not work, do we expect this loop to happen at least 2-3x?
Could we move the sleep to the beginning of the loop body to increase likelihood that this loop exits on an earlier iteration? Hopefully save API calls we don't expect to yield a successful get on the program run.

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: I'm not sure this will have the behavior you expect.

what happens when you do a get on the program run id that isn't present yet? 404 or empty body?

I personally get 404 on a 6.1.1 instance for a random uuid i generated. Has the API behavior changed?

Screenshot 2020-05-22 at 9 58 30 AM

Comment (jaketf) [reply airflow/providers/google/cloud/hooks/datafusion.py]: FYI you can get to this convenient UI by clicking system admin > configuration > Make HTTP calls
It's very useful for getting used to / testing the CDAP REST API.

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/datafusion.py]: This CDAP API is very convoluted.

Seems like there are [several program types](https://docs.cask.co/cdap/6.1.1/en/reference-manual/http-restful-api/lifecycle.html#H3581) and this is hard coding `workflows` and `DataPipelineWorkflow` .

I think `DataPipelineWorkflow` this will not be present for streaming pipelines instead you have to poll a `spark` program.

I'm not sure how many other scenarios require other program types.
It would be good to get someone from CDAP community to review this.

Screenshot 2020-05-22 at 10 21 00 AM

Screenshot 2020-05-22 at 10 20 27 AM

Comment (jaketf) [reply airflow/providers/google/cloud/hooks/datafusion.py]: CC: @sreevatsanraman

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/datafusion.py]: You are requesting `.../runs/run-id`, the code here is calling `.../runs` to get list of all runs because we don't know yet the proper CDAP `run-id`. I assume that this request should be successful unless something wrong is with API / network.

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/datafusion.py]: This seems to be a bigger change because we will have to adjust each method that uses `DataPipelineWorkflow` in URI. So, I would say we can do this but in a follow up PR.

Btw. I was relying on Google docs: https://cloud.google.com/data-fusion/docs/reference/cdap-reference#start_a_batch_pipeline

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/datafusion.py]: Usually it's successful on 2nd loop but sometimes I get to the 3rd. I will move the `sleep` to the beginning as you suggested so it should decrease number of requests.

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/datafusion.py]: Done

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [reply airflow/providers/google/cloud/hooks/datafusion.py]: Ah I see, my mistake. 
You'll get 200 and empty collection if there's no runs.

Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/datafusion.py]: Yes, so I will retry the request. I am not sure if there's anything we can do about this. When we call this method we are expecting to see some runs

Comment (sreevatsanraman) [reply airflow/providers/google/cloud/hooks/datafusion.py]: @jaketf the API that CDAP exposes is the basic building blocks of programs. Which are workflows, spark, mapreduce jobs etc. The Data Fusion pipelines use workflows for batch jobs and spark streaming jobs for realtime. The operators should wait for the batch jobs and not wait for the streaming ones.

Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/datafusion.py]: @sreevatsanraman how can we distinguish those two types? According to Data Fusion CDAP API reference users should use the same endpoint to start both batch and streaming pipelines:
https://cloud.google.com/data-fusion/docs/reference/cdap-reference#start_a_batch_pipeline

**Comment by turbaszek** [comment]
Hi @jaketf @sreevatsanraman what should we do to move this forward?

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [reply airflow/providers/google/cloud/hooks/datafusion.py]: I think we should just handle batch pipelines in this PR (as this is implicitly all the current operator does). Also, anecdotally, I think this covers 90% of use cases for airflow. In the field i have not see a lot of streaming orchestration with airflow.

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (jaketf) [reply airflow/providers/google/cloud/hooks/datafusion.py]: As long as we cover batch pipelines (with spark or MR backend I think we should be good)

**Comment by jaketf** [comment]
Thanks for following up @turbaszek.
*tl;dr* I think we should merge this PR as it fixes the immediate issue. We can file a lower priority issue to handle streaming pipelines in the future. This can be an additional kwarg that accepts a streaming flag and uses a different paths for polling. 

I've updated the threads. I agree I think we should keep this PR small and focused on patching the existing operator for starting data fusion batch pipelines.

In general I think batch is more used than streaming and spark is more used than MR.
In batch both MR and spark can be polled at the .../DataPipelineWorkflow/runs/run_id endpoint.

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (rachael-ds) [reply airflow/providers/google/cloud/hooks/datafusion.py]: You can avoid using the faux run Id by making a call to the batch start pipeline endpoint - the run id will be returned in this case. For example:
TYPE: POST
URL: 'https://xxx.datafusion.googleusercontent.com/api/v3/namespaces/default/start'
BODY: 
[{"appId": "app_id", "programType": "workflow", "programId": "DataPipelineWorkflow","runtimeargs": {}}]

Batch start pipeline endpoint info:
https://docs.cdap.io/cdap/current/en/reference-manual/http-restful-api/lifecycle.html#H3293 (documentation does not currently reflect that the run Id is returned)

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (mik-laj) [inline airflow/providers/google/cloud/hooks/datafusion.py]: Should we URL encode this parameter?

Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/datafusion.py]: Done

File: airflow/providers/google/cloud/operators/datafusion.py
Comment (rachael-ds) [inline airflow/providers/google/cloud/operators/datafusion.py:662]: If the success state is COMPLETED, we will timeout before the pipeline run completes.
It can take > 5 minutes to just provision a Data Fusion pipeline run. Some pipelines can take hours to complete.
Can we increase the default timeout to 1 hour?

File: airflow/providers/google/cloud/operators/datafusion.py
Comment (rachael-ds) [inline airflow/providers/google/cloud/operators/datafusion.py:630]: missing info for new _pipeline_timeout_ parameter

File: airflow/providers/google/cloud/hooks/datafusion.py
Comment (rachael-ds) [inline airflow/providers/google/cloud/hooks/datafusion.py:485]: Just an FYI - this is the API request to start multiple pipelines. 
There will eventually be a fix return the run Id as part of the API request to run a _single_ pipeline. 
We can revert to your original URL when this is available. For context:
https://issues.cask.co/browse/CDAP-7641

**Review by mik-laj — APPROVED** [review]

File: airflow/providers/google/cloud/operators/datafusion.py
Comment (jaketf) [reply airflow/providers/google/cloud/operators/datafusion.py:662]: The contract of this operator is to start a pipeline. not wait til pipeline completion.
10 mins is reasonable timeout.
COMPLETED is just a success state in case it's a super quick pipeline that completes between polls.
We can add a sensor for waiting on pipeline completion (which should use reschedule mode if it expects to be so long).

Comment (rachael-ds) [reply airflow/providers/google/cloud/operators/datafusion.py:662]: As part of these changes you can now pass in a parameter to have the operator wait for pipeline completion (not just pipeline start).
sensor + reschedule mode sounds like a good suggestion, thanks

Comment (turbaszek) [reply airflow/providers/google/cloud/operators/datafusion.py:662]: I've created an issue to limit scope of this PR
https://github.com/apache/airflow/issues/9300

**Review by mik-laj — APPROVED** [review]

**Pull request merged**
